text/0003-performance-data.md

```diff
@@ -0,0 +1,83 @@
+- Start Date: 1-29-2021
+- RFC PR: [Unity-Technologies/com.unity.multiplayer.rfcs#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Currently there is Transport performance data and MLAPI performance data that is not being propagated up externally for any consumer to render or consume. The use cases for this would be useful for technologies such as a **Network Stats Overlay** or a **Profiler** (**Network Messages** and **Network Operation** need this data in the vanilla Unity Profiler). Users need a viable and low overhead way to obtain this data (such as RPCs sent this frame/tick) from MLAPI and utilize this data in the way that they want for their own metrics.
+
+# Motivation
+[motivation]: #motivation
+
+This is needed to provide both a default and user specific implementations for future Network Profilers and Network Stats overlays. The MLAPI source code can remain Open Source, and yet, the data being streamed can be hooked up to external and low level systems.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to allow first and third party rendering of the data we will need to allow both MLAPI and its transports to propagate data up the chain. 
+If the profiler flags are enabled the following will need to happen: 
+Transports will need to output the number of bytes being sent in the current tick and the latency. The data will be written out to a ProfilerTickData chunk. This data will be generic and transport agnostic.
+MLAPI will need to output the number of networked objects and the amount of RPCs being sent. This data will be added to the ProfilerTickData chunk.
+
+External Tools and APIs will be able to then get this fully written ProfilerTickData at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a struct like object, accessing its data will be as simple as accessing fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in ```NetworkManager.LateUpdate()``` for the MLAPI and ```Transport.Send()``` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data).
+
+```cs
+class ProfilerTickData {
```

Comment:
We should make sure that this is done in such a way where no additional allocations are needed to gather this data, either by re-using the same object each frame or by using structs. This would inform the technical approach to gathering the data

Reply:
since it's a relatively small container, we could make it a `struct` and forget about heap pressure.
on the other hand, we could leave it as a `class`, cache it internally on the producer-side (transport in this case) and re-use that existing reference to modify properties everytime we call transport's (let's say) `TickProfilerStats()` API.